col_plan vignette (#114)
GSK-Biostatistics/tlang: vignettes/col_plan.Rmd

Let's look at each of these key workflows in more detail.

# Option 1: A single `column` variable 

Let's consider the following dataset which includes a single `column` variable:

```{r}
dat <- tribble(
    ~group,     ~label,  ~my_col,    ~parm, ~val,
    "g1", "rowlabel1",  "col1"  ,  "value",    1,
    "g1", "rowlabel1",  "col2"  ,  "value",    1,
    "g1", "rowlabel1",  "mycol3",  "value",    1,
    "g1", "rowlabel1",  "col4"  ,  "value",    1,
    "g1", "rowlabel1",  "mycol5",  "value",    1,
    "g1", "rowlabel2",  "col1"  ,  "value",    2,
    "g1", "rowlabel2",  "col2"  ,  "value",    2,
    "g1", "rowlabel2",  "mycol3",  "value",    2,
    "g1", "rowlabel2",  "col4"  ,  "value",    2,
    "g1", "rowlabel2",  "mycol5",  "value",    2,
    "g2", "rowlabel3",  "col1"  ,  "value",    3,
    "g2", "rowlabel3",  "col2"  ,  "value",    3,
    "g2", "rowlabel3",  "mycol3",  "value",    3,
    "g2", "rowlabel3",  "col4"  ,  "value",    3,
    "g2", "rowlabel3",  "mycol5",  "value",    3)
  
```

> ## No column spanners

Reviewer: I think a title that is more descriptive of what it is, rather than what it isnt may be more helpful for folks.
```suggestion
## Simple Column Selection
```

In the case of a single column variable with no column spanners, `col_plan` behaves identically to `dplyr::select`. 

Let's say we want to remove one of the columns:
```{r}
tfrmt(
    group = group,
    label = label,
    param = parm,
    values = val,
    column = my_col,
    body_plan = body_plan(
      frmt_structure(group_val = ".default", label_val = ".default", frmt("x"))
    ),
    col_plan = col_plan(
      -mycol5            
    )
  ) %>% 
  print_to_gt(dat)
```

Just like `dplyr::select`, we can also reorder, rename, and remove columns, using `tidyselect` syntax if we'd like:

```{r}
tfrmt(
    group = group,
    label = label,
    param = parm,
    values = val,
    column = my_col,
    body_plan = body_plan(
      frmt_structure(group_val = ".default", label_val = ".default", frmt("x"))
    ),
    col_plan = col_plan(
      group, 
      label, 

>       starts_with("col"),

Reviewer: Should we default to keeping these values when they are defined? I am feeling as though the users shouldn't need to define them. what are everyone's thoughts?

Author: I tend to agree. I like that people have control over order/dropping them, but i think it would be a nice default to keep them in if not specified.

      new_col_3 = mycol3,  
      -mycol5            
    )
  ) %>% 
  print_to_gt(dat)
```

## Column spanners

>  The `col_plan` accepts a combination of column names and/or span structure (`span_structure`) objects. Recall that `*_structure` objects are for modifying the *contents* (i.e., text) of the table; in the `col_plan`, we use `span_structure`s to add column spanners. These are placed strategically in the `col_plan` to inform `tlang` of the hierarchy of spanners vs. columns.

Reviewer: ```suggestion
 The `col_plan` accepts a combination of tidyselect expressions, column names, and/or span structure (`span_structure`) objects. Recall that `*_structure` objects are for modifying the *contents* (i.e., text) of the table; in the `col_plan`, we use `span_structure`s to add column spanners. These are placed strategically in the `col_plan` to inform `tlang` of the hierarchy of spanners and columns.
 
A `span_structure` is constructed from a label which defines the spanner text, and a combination of tidyselect expressions and column names.
```

 
Let's try putting a spanner over the columns starting with "col":

```{r}
tfrmt(
    group = group,
    label = label,
    param = parm,
    values = val,
    column = my_col,
    body_plan = body_plan(
      frmt_structure(group_val = ".default", label_val = ".default", frmt("x"))
    ),
    col_plan = col_plan(
      group, 
      label, 
      span_structure(
        "column cols",
        starts_with("col")
        ),
      new_col_3 = mycol3,  
     - mycol5            
    )
  ) %>% 
  print_to_gt(dat)
```

> We can also do any number of hierarchical spanners:

Reviewer: ```suggestion
A `span_structure` can also contain one or more child `span_structure` entries, defining a hierarchy of spanners:
```

    "g1", "rowlabel2",             NA,        NA , "mycol3",  "value",    2,
    "g1", "rowlabel2",  "column cols", "col 4"   , "col4"  ,  "value",    2,
    "g1", "rowlabel2",             NA,         NA, "mycol5",  "value",    2,
    "g2", "rowlabel3",  "column cols", "cols 1,2", "col1"  ,  "value",    3,
    "g2", "rowlabel3",  "column cols", "cols 1,2", "col2"  ,  "value",    3,
    "g2", "rowlabel3",             NA,         NA, "mycol3",  "value",    3,
    "g2", "rowlabel3",  "column cols", "col 4"   , "col4"  ,  "value",    3,
    "g2", "rowlabel3",             NA,         NA, "mycol5",  "value",    3,
  )
```

Notice that our `col_plan` is considerably simpler and our higher level columns are repurposed as column spanners:

```{r}
tfrmt(
    group = group,
    label = label,
    param = parm,
    values = val,
    column = c(span2, span1, my_col),
    body_plan = body_plan(
      frmt_structure(group_val = ".default", label_val = ".default", frmt("x"))
    ),
    col_plan = col_plan(
      group, 
      label, 
      starts_with("col"),
      new_col_3 = mycol3,  
      -mycol5            
    )
  ) %>% 
  print_to_gt(dat)
```

> When there are multiple `column` variables provided, `tlang` concatenates them behind the scenes using a special delimiter. Therefore, the underlying column names in the final table are a combination of the spanner(s) and the lowest level column label. In order to change the order of spanning groups, we may use`contains()` to reference any portion of those names. 

Reviewer: Not sure if we want to share the details on how `tlang` behaves, but understand why to make it so folks understand how they can use `contains()`. 

```suggestion
When there are multiple `column` variables provided, `tlang` concatenates them behind the scenes using a special delimiter. Therefore, the underlying column names in the final table are a combination of the spanner(s) and the lowest level column label. In order to change the order of spanning groups, we may use `contains()` to reference any portion of those names. 
```

Author: I'm going to revise so it's somewhere in the middle.

> Reordering with multiple `column` variables is simplest when the lowest level column variable contains unique values (i.e., there is 1 value per column in the final table), like in the example above. Let's suppose we want to move the "col 4" column to the beginning and also reorder "col1" and "col2". To achieve this, we can just select "col 4" and then "col2" and finally all the other columns. 

Reviewer: ```suggestion
Reordering with multiple `column` variables is simplest when the lowest level column variable contains unique values (i.e., there is 1 value per column in the final table), like in the example above. Let's suppose we want to move the "col 4" column to the beginning and also reorder "col1" and "col2". To achieve this, we can select "col 4" and then "col2" and finally all the other columns. 
```
